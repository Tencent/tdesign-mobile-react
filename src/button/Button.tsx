import React, { useMemo } from 'react';
import classnames from 'classnames';
import TLoading from '../loading';
import parseTNode from '../_util/parseTNode';
import { TdButtonProps } from './type';
import { buttonDefaultProps } from './defaultProps';
import { usePrefixClass } from '../hooks/useClass';
import useDefaultProps from '../hooks/useDefaultProps';
import useHover from '../hooks/useHover';

export interface ButtonProps
  extends TdButtonProps,
    Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'content' | 'children'> {}

const Button: React.FC<ButtonProps> = (originProps) => {
  const props = useDefaultProps(originProps, buttonDefaultProps);
  const {
    className,
    style,
    block,
    children,
    content,
    disabled,
    ghost,
    icon,
    suffix,
    loading,
    shape,
    size,
    theme,
    type,
    variant,
    onClick,
    loadingProps,
  } = props;
  const buttonClass = usePrefixClass('button');
  const childNode = content || children;

  const hoverDisabled = useMemo(() => disabled || loading, [disabled, loading]);
  const ref = useHover({ className: `${buttonClass}--hover`, disabled: hoverDisabled });

  return (
    <button
      ref={ref}
      type={type}
      className={classnames(
        [
          `${buttonClass}`,
          `${buttonClass}--size-${size}`,
          `${buttonClass}--${variant}`,
          `${buttonClass}--${theme}`,
          `${buttonClass}--${shape}`,
          className,
        ],
        {
          [`${buttonClass}--ghost`]: ghost,
          [`${buttonClass}--loading`]: loading,
          [`${buttonClass}--disabled`]: disabled,
          [`${buttonClass}--block`]: block,
        },
      )}
      style={style}
      onClick={!loading && !disabled ? onClick : undefined}
      disabled={disabled || loading}
    >
      {loading ? <TLoading inheritColor {...loadingProps} /> : parseTNode(icon)}
      {childNode && <span className={`${buttonClass}__content`}> {parseTNode(childNode)}</span>}
      {parseTNode(suffix)}
    </button>
  );
};

Button.displayName = 'Button';

export default Button;
