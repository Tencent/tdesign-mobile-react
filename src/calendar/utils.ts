import { TCalendarValue } from './type';

export function getMonthByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setMonth(innerDate.getMonth() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export function getYearByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setFullYear(innerDate.getFullYear() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export const getPrevMonth = (date: TCalendarValue) => getMonthByOffset(date, -1);
export const getNextMonth = (date: TCalendarValue) => getMonthByOffset(date, 1);
export const getPrevYear = (date: TCalendarValue) => getYearByOffset(date, -1);
export const getNextYear = (date: TCalendarValue) => getYearByOffset(date, 1);
