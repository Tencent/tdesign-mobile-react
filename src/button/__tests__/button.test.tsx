import React from 'react';
import { describe, it, expect, render, vi, fireEvent } from '@test/utils';

import { AppIcon } from 'tdesign-icons-react';
import Button from '../Button';

const prefix = 't';
const name = `.${prefix}-button`;
const buttonText = '按钮组件';

describe('Button', () => {
  describe('props', () => {
    it(': block', () => {
      const { container } = render(<Button block />);
      expect(container.querySelector(`${name}--block`)).toBeTruthy();
    });
    it(': children', () => {
      const { queryByText } = render(<Button>{buttonText}</Button>);
      expect(queryByText(buttonText)).toBeInTheDocument();
    });
    it(': content', () => {
      const { queryByText } = render(<Button content={buttonText} />);
      expect(queryByText(buttonText)).toBeInTheDocument();
    });
    it(': disabled', () => {
      const { container } = render(<Button disabled />);
      expect(container.querySelector(`${name}--disabled`)).toBeTruthy();
    });
    it(': ghost', () => {
      const { container } = render(<Button ghost />);
      expect(container.querySelector(`${name}--ghost`)).toBeTruthy();
    });
    it(': icon', () => {
      const { container } = render(<Button icon={<AppIcon />} />);
      expect(container.querySelector('.t-icon-app')).toBeTruthy();
    });
    it(': loading', () => {
      const { container } = render(<Button loading />);
      expect(container.querySelector(`${name}--loading`)).toBeTruthy();
    });
    it(': loadingProps', () => {
      const { container } = render(<Button loading loadingProps={{ duration: 1000 }} />);
      expect(container.querySelector('.t-loading')).toBeInTheDocument();
    });
    it(': shape', () => {
      const shapes = ['rectangle', 'square', 'round', 'circle'] as const;
      shapes.forEach((shape) => {
        const { container } = render(<Button shape={shape} />);
        expect(container.querySelector(`${name}--${shape}`)).toBeTruthy();
      });
    });
    it(': size', () => {
      const sizes = ['extra-small', 'small', 'medium', 'large'] as const;
      sizes.forEach((size) => {
        const { container } = render(<Button size={size} />);
        expect(container.querySelector(`${name}--size-${size}`)).toBeTruthy();
      });
    });
    it(': suffix', () => {
      const { container } = render(<Button suffix={<AppIcon />} />);
      expect(container.querySelector('.t-icon-app')).toBeInTheDocument();
    });
    it(': theme', () => {
      const themes = ['default', 'primary', 'danger', 'light'] as const;
      themes.forEach((theme) => {
        const { container } = render(<Button theme={theme} />);
        expect(container.querySelector(`${name}--${theme}`)).toBeTruthy();
      });
    });
    it(': variant', () => {
      const variants = ['base', 'outline', 'dashed', 'text'] as const;
      variants.forEach((variant) => {
        const { container } = render(<Button variant={variant} />);
        expect(container.querySelector(`${name}--${variant}`)).toBeTruthy();
      });
    });
  });

  describe('events', () => {
    it(': onClick', () => {
      const handleClick = vi.fn();
      const { container } = render(<Button onClick={handleClick} />);
      fireEvent.click(container.firstChild);
      expect(handleClick).toHaveBeenCalled();
    });
    it(': onClick disabled', () => {
      const handleClick = vi.fn();
      const { container } = render(<Button disabled onClick={handleClick} />);
      fireEvent.click(container.firstChild);
      expect(handleClick).not.toHaveBeenCalled();
    });
    it(': onClick loading', () => {
      const handleClick = vi.fn();
      const { container } = render(<Button loading onClick={handleClick} />);
      fireEvent.click(container.firstChild);
      expect(handleClick).not.toHaveBeenCalled();
    });
  });
});
