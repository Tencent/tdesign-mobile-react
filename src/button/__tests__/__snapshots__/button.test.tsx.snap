// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > content 1`] = `
<span
  class="t-button__content"
>
   
  按钮组件
</span>
`;

exports[`Button 组件测试 > content 1`] = `
<span
  class="t-button__content"
>
   
  按钮组件
</span>
`;
